test/utils/floating_point_utils.hpp

```diff
@@ -0,0 +1,69 @@
+// Fizzy: A fast WebAssembly interpreter
+// Copyright 2020 The Fizzy Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <cmath>
+#include <cstdint>
+#include <limits>
+#include <type_traits>
+
+namespace fizzy::test
+{
+template <typename T, typename = std::enable_if<std::is_floating_point_v<T>>>
+struct FPUtils
+{
+    using Limits = std::numeric_limits<T>;
+
+    static constexpr auto NumMantissaBits = Limits::digits - 1;
+    static constexpr auto MantissaMask = (uint64_t{1} << NumMantissaBits) - 1;
+
+    static constexpr auto CanonicalNaNPayload = uint64_t{1} << (NumMantissaBits - 1);
+};
+
+template <typename T, typename = std::enable_if<std::is_floating_point_v<T>>>
+uint64_t as_uint64(T value) noexcept
+{
+    constexpr auto num_bytes = sizeof(T);
+    uint8_t bytes[num_bytes];
+    __builtin_memcpy(bytes, &value, num_bytes);
+
+    uint64_t u = 0;
+    for (size_t i = 0; i < num_bytes; ++i)
+        u |= uint64_t{bytes[i]} << (i * 8);  // little-endian
+    return u;
+}
+
+template <typename T, typename = std::enable_if<std::is_floating_point_v<T>>>
+T from_uint64(uint64_t value) noexcept
+{
+    constexpr auto num_bytes = sizeof(T);
+    uint8_t bytes[num_bytes];
+
+    for (size_t i = 0; i < num_bytes; ++i)
+        bytes[i] = static_cast<uint8_t>(value >> (i * 8));  // little-endian
+
+    T result;
+    __builtin_memcpy(&result, bytes, num_bytes);
+    return result;
+}
+
+template <typename T, typename = std::enable_if<std::is_floating_point_v<T>>>
+uint64_t get_nan_payload(T value) noexcept
+{
+    if (!std::isnan(value))
+        return 0;  // NaN payload is never 0.
+
+    return as_uint64(value) & FPUtils<T>::MantissaMask;
+}
+
+template <typename T, typename = std::enable_if<std::is_floating_point_v<T>>>
+T nan(uint64_t payload = FPUtils<T>::CanonicalNaNPayload) noexcept
```

Comment:
Couldn't we use `std::nan`/`std::nanf` for this?

Reply:
True, but they do it in an implementation-defined manner related to printing NaN "values". Maybe there is a way to find a combination of chars that gives wasm canonical and arithmetic NaNs, but I decided to go with my own NaN inspection helpers. Also to learn the internal structure of floats.

Reply:
Added some unit tests comparing with `std::nan()`. It produces only quiet (aka arithmetic) nans. I prefer testing others (i.e. signaling) nans too.